src/primitives/primitive_style.rs

```diff
@@ -118,7 +126,11 @@ where
     /// Returns the fill area.
     pub(in crate::primitives) fn fill_area<P: OffsetOutline>(&self, primitive: &P) -> P {
         // saturate offset at i32::min_value() if stroke width is to large
-        let offset = -self.inside_stroke_width().saturating_as::<i32>();
+        let offset = if self.stroke_style == StrokeStyle::Solid {
+            -self.inside_stroke_width().saturating_as::<i32>()
+        } else {
+            0
```

Comment:
```suggestion
            // do not shrink the fill area for non solid borders, because the entire fill
            // area is visible through the gaps in the border            
            0
```